hips/hip-9999.md

```diff
@@ -0,0 +1,220 @@
+Changes can be breaking CRD scheme, incompatability with their version of Kubernetes or infrastructure, performance degregation, or any change the cusumer doesn't want.
+
+Chart publishers need to publish breaking changes because of Kubernetes API changes or CRD updates to remediate design flaws or critical implementation bugs.
+Users can specify a `--version` when creating a release or performing an upgrade, but it is not possible to centrally know which releases need a specific version or what that version is.
+
+Chart authors can add deprecation messages to their charts (e.g. [gitlab does this](https://docs.gitlab.com/charts/architecture/decisions.html#breaking-changes-via-deprecation)) but it doesn't stop users from upgrading to releases without deprecation messages.
+Chart authors can also publish breaking chart versions to new repositories but this causes more consumer pain points to discover the repositories when they do want to upgrade.
+
+Chart authors should also use the `kubeVersion` field to specify compatible Kubernetes versions, but this may not be enough to avoid breaking changes based on other resources and controllers is the cluster.
+
+## Rationale
+
+If helm is "the package manager for Kubernetes" it should have similar package manager features that users are used to such as `apt` and `yum`.
+With both of those package managers there is a centralized way to specify a version of a package and keep that version installed even when a user runs an `upgrade` command against a release.
+Centralizing the desired version guarantees that all users, automation, and libraries that use the package manager also honor the version locks in place.
+
+Making something native in helm that supports locking will be critical as users consume helm charts via other tools and workflows (e.g. gitops via `flux`).
+It will reduce errors because `helm` commands can be run from any users machine with access to the cluster.
+This is opposed to a helm plugin model where a user or tool may not have the plugin installed.
+
+Helm releases do not have the same constraints has traditional package managers.
+A cluster can have multiple versions of a chart deployed in different namespaces and each release can have various dependencies.
+Because releases are self-contained on a per-release basis, version locking information should also exist within the release metadata and be read and processed during template rendering.
+
+## Specification
+
+To implement version locking on a release an optional field should be added to the release secret that specifies a desired version.
+
+```
+apiVersion: v1                                                           
+data:                                                                 
+  release: SDRzSUF...
+kind: Secret
+metadata:
+  creationTimestamp: "2021-02-23T18:07:25Z"
+  desiredChartVersion: "8.0.*"             <---- this field is added
```

Comment:
The field `desiredChartVersion` is on the secret and we can't have it here. The release data can be stored in secrets, configmaps, and SQL. Release information is pluggable. Just this week I was talking with someone about a new place to store release information and I know some folks are using the SQL storage back-end. See the drivers.

It should be on the release object. But, this brings in issues when older versions of Helm v3 try to work with the release object... they won't know about the new field to respect it. How would that work or does this need to wait for Helm v4?

Reply:
I'm ok if this change doesn't make it into helm 3. It would be nice but I am not familiar with all of the backwards compatibility concerns. I also wasn't sure where exactly data like this should be placed in the release. If it is more suited to be put in the secret that is fine too. I assumed putting it in the metadata would make it easier to view if a release is locked outside of helm tooling (e.g. `kubectl`)